src/RequestParser.cpp

```diff
@@ -0,0 +1,551 @@
+ * @throws std::runtime_error If an invalid character is encountered in the query,
+ *         or if another '?' is found, an error is thrown and the error code is set to 400.
+ */
+void	RequestParser::parseUriQuery(const std::string& requestLine, int& index)
+{
+	while (requestLine[++index]) {
+		if (requestLine[index] == ' ' || requestLine[index] == '#') {
+			index--;
+			break;
+		}
+		else if (!isValidURIChar(requestLine[index]) || requestLine[index] == '?') {
+			m_errorCode = 400;
+			throw std::runtime_error(ERR_URI_INVALID_CHAR);
+		}
+		else
+			m_request.uri.query.push_back(requestLine[index]);
+	}
+}
+
+/**
+ * @brief Parses the fragment component of the URI from the request line.
+ *
+ * This function extracts the fragment component of the URI from the provided request line.
+ * It starts parsing at the given index and continues until a space character is encountered.
+ * Each character in the fragment is validated using `isValidURIChar`.
+ * If an invalid character or another '#' is encountered,
+ * an error code is set, and a `std::runtime_error` is thrown.
+ *
+ * @param requestLine The request line string from which the fragment component of the URI is to be parsed.
+ * @param index The current index in the request line string where the fragment component starts.
+ *              This index is updated as the fragment is parsed.
+ * @throws std::runtime_error If an invalid character is encountered in the fragment,
+ *         or if another '#' is found, an error is thrown and the error code is set to 400.
+ */
+void	RequestParser::parseUriFragment(const std::string& requestLine, int& index)
+{
+	while (requestLine[++index]) {
+		if (requestLine[index] == ' ') {
+			index--;
+			break;
+		}
+		else if (!isValidURIChar(requestLine[index]) || requestLine[index] == '#') {
+			m_errorCode = 400;
+			throw std::runtime_error(ERR_URI_INVALID_CHAR);
+		}
+		else
+			m_request.uri.fragment.push_back(requestLine[index]);
+	}
+}
+
+/**
+ * @brief Parses the HTTP version from the request line.
+ *
+ * This function extracts and validates the HTTP version from the provided request line.
+ * It checks if the version format starts with "HTTP/",
+ * followed by a valid major and minor version number.
+ * If any format error is detected (e.g., missing digits or incorrect delimiter),
+ * an appropriate error code is set, and a `std::runtime_error` is thrown.
+ *
+ * @param requestLine The request line string from which the HTTP version is to be parsed.
+ * @return A substring of `requestLine` starting from the character after the parsed version.
+ * @throws std::runtime_error If the HTTP version format is invalid,
+ *         an error is thrown and the error code is set to 400.
+ */
+std::string	RequestParser::parseVersion(const std::string& requestLine)
+{
+	if (requestLine.substr(0, 5) != "HTTP/") {
+		m_errorCode = 400;
+		throw std::runtime_error(ERR_INVALID_VERSION_FORMAT);
+	}
+	int	i = 5;
+	if (!isdigit(requestLine[i])) {
```

Comment:
Here would I recommend to also check the value of the digit.

Currently in your tests HTTP/2.0 is valid. I compared it with NGINX and it did not support this version.
Besides that I would also say, that we only support HTTP/1.1 (maybe also HTTP/1.0 but NGINX also did not accept that)

Reply:
I agree with Chris and would have the parser verify whether the requested version is supported or not. In case of an unsupported version the error code is set to 505 and then the response builder should create an appropriate message (see https://http.dev/505).

This approach is in alignment with the general implementation that the parser identifies errors and assigns the corresponding error code which in turn is used by the response builder to generate an answer.

Reply:
I disagree that it's the parsers job to identify whether the version is supported. The parser identifies formal errors, e.g. there is a letter, when there should be a number. 
The response builder will also set error codes for himself, which the parser can't identify. For example if the requested resource is not available. 

However since we most certainly won't maintain this project and implement more http version, I'm fine either way.

Reply:
While it is already looking at the version, it may check for the support as well. That way we catch a unsupported version early on and save us all the remaining superfluous parsing and following work which would be executed for nil. 

As a reference: when you take a look at the go language implementation, their implementation of the request reader (parser) also does the version check: 
https://github.com/golang/go/blob/773767def0e0f29584a69bd760430167b7479d7d/src/net/http/request.go#L826

Reply:
I think Sirui has a good point that the check for the http version can be implemented in the config file parser and therefore checks earlier if it is invalid. 

I will implement this in the parsing process

Reply:
That's a hard to beat argument :D 
Agreed!